File: android/src/main/kotlin/fl/channel/FlChannelPlugin.kt
package fl.channel


import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlChannelPlugin */
class FlChannelPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var plugin: FlutterPlugin.FlutterPluginBinding

    companion object {
        private var eventChannels: MutableMap<String, FlEventChannel> = mutableMapOf()
        fun getEventChannel(name: String): FlEventChannel? {
            return eventChannels[name]
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        plugin = binding
        channel = MethodChannel(binding.binaryMessenger, "fl_channel")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "create" -> {
                val name = call.arguments as String
                if (!eventChannels.containsKey(name)) {
                    val eventChannel = FlEventChannel(name, plugin.binaryMessenger)
                    eventChannels[name] = eventChannel
                }
                result.success(true)
            }

            "sendEventFromNative" -> {
                val args = call.arguments as Map<*, *>
                val name = args["name"] as String;
                val data = args["data"] as Any;
                result.success(getEventChannel(name)?.send(data) ?: false)
            }

            "dispose" -> {
                val name = call.arguments as String
                if (eventChannels.containsKey(name)) {
                    getEventChannel(name)?.cancel()
                    eventChannels.remove(name)
                }
                result.success(true)
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

}

File: ios/Classes/FlChannelPlugin.swift
import Flutter
import UIKit

public class FlChannelPlugin: NSObject, FlutterPlugin {
    var channel: FlutterMethodChannel
    var messenger: FlutterBinaryMessenger

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "fl_channel", binaryMessenger: registrar.messenger())
        let instance = FlChannelPlugin(registrar.messenger(), channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public init(_ messenger: FlutterBinaryMessenger, _ channel: FlutterMethodChannel) {
        self.channel = channel
        self.messenger = messenger
        super.init()
    }

    private static var eventChannels: [String: FlEventChannel] = [:]

    public static func getEventChannel(name: String) -> FlEventChannel? {
        if eventChannels.keys.contains(name) {
            return eventChannels[name]
        }
        return nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "create":
            var name = call.arguments as! String
            if !FlChannelPlugin.eventChannels.keys.contains(name) {
                var eventChannel = FlEventChannel(name, messenger)
                FlChannelPlugin.eventChannels[name] = eventChannel
            }
            result(true)
        case "sendEventFromNative":
            var args = call.arguments as! [String: Any]
            var name = args["name"] as! String
            var data = args["data"]
            var state = FlChannelPlugin.getEventChannel(name: name)?.send(data)
            result(state ?? false)
        case "dispose":
            var name = call.arguments as! String
            if FlChannelPlugin.eventChannels.keys.contains(name) {
                FlChannelPlugin.eventChannels[name]?.cancel()
                FlChannelPlugin.eventChannels.removeValue(forKey: name)
            }
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        channel.setMethodCallHandler(nil)
    }
}
